libraries/botbuilder-runtime/src/coreBotAdapter.ts

```diff
@@ -1,26 +1,86 @@
+    constructor(
+        authConfig: AuthenticationConfiguration,
+        private readonly conversationState: ConversationState,
+        userState: UserState
+    ) {
         super({ authConfig });
 
         // attach storage?
         useBotState(this, userState, conversationState);
 
         this.onTurnError = async (turnContext, err) => {
-            console.error(err);
-
-            await turnContext.sendActivity(err.message, err.message);
+            console.error('[onTurnError] unhandled error', err);
 
-            try {
-                await conversationState.delete(turnContext);
-            } catch (err) {
-                console.error('Exception caught on attempting to delete conversation state', err);
-            }
+            await this.sendErrorMessage(turnContext, err);
+            await this.sendEoCToParentIfSkill(turnContext, err);
+            await this.clearConversationState(turnContext);
         };
     }
+
+    private async sendErrorMessage(turnContext: TurnContext, error: Error): Promise<void> {
+        try {
+            let errorMessageText = 'The bot encountered an error or bug';
+            let errorMessage = MessageFactory.text(errorMessageText, errorMessageText, InputHints.IgnoringInput);
+            await turnContext.sendActivity(errorMessage);
+
+            errorMessageText = 'To continue to run this bot, please fix the bot source code.';
+            errorMessage = MessageFactory.text(errorMessageText, errorMessageText, InputHints.ExpectingInput);
+            await turnContext.sendActivity(errorMessage);
+
+            await turnContext.sendTraceActivity(
+                'OnTurnError Trace',
+                error,
+                'https://www.botframework.com/schemas/error',
+                'TurnError'
+            );
+        } catch (err) {
+            console.error('Exception caught in sendErrorMessage', err);
+        }
+    }
+
+    private async sendEoCToParentIfSkill(turnContext: TurnContext, error: Error): Promise<void> {
+        if (!this.isSkillBot(turnContext)) {
+            return;
+        }
+
+        try {
+            const endOfConversation = ActivityEx.createEndOfConversationActivity();
+            endOfConversation.code = 'SkillError';
+            endOfConversation.text = error.message;
+            await turnContext.sendActivity(endOfConversation);
+        } catch (err) {
+            console.error('Exception caught in sendEoCToParentIfSkill', err);
+        }
+    }
+
+    private async clearConversationState(turnContext: TurnContext): Promise<void> {
+        try {
+            await this.conversationState.delete(turnContext);
+        } catch (err) {
+            console.error('Exception caught in clearConversationState', err);
+        }
+    }
+
+    private isSkillBot(turnContext: TurnContext): boolean {
+        const claimsIdentity = turnContext.turnState.get<ClaimsIdentity>(BotAdapter.BotIdentityKey);
```

Comment:
Tears of joy at `.get<ClaimsIdentity>`()!

Reply:
If you're using the static member here, we need to change the the rest of the SDK where it uses `turnContext.adapter.BotIdentityKey` to set and get the ClaimsIdentity in case a developer subclassed the old BotAdapter and overrode the BotIdentityKey. Is this a correct observation?

Reply:
Oi - good catch. I'd rather not change `BotAdapter`, how would I go about fetching the adapter from turn context?

Reply:
It's straight forward, just [`turnContext.adapter`](https://github.com/microsoft/botbuilder-js/blob/a9003e59ea9df7795c2c6d1dcb660ccdb1295775/libraries/botbuilder-core/src/turnContext.ts#L725-L727) which returns a BotAdapter 👍🏼

Reply:
Wonderful, thanks!